src/core/desktopentry.cpp

```diff
@@ -407,7 +438,138 @@ DesktopEntry* DesktopEntryManager::heuristicLookup(const QString& name) {
 
 ObjectModel<DesktopEntry>* DesktopEntryManager::applications() { return &this->mApplications; }
 
-DesktopEntries::DesktopEntries() { DesktopEntryManager::instance(); }
+void DesktopEntryManager::handleFileChanges() {
+	qCDebug(logDesktopEntry) << "Directory change detected, performing full rescan";
+
+	if (this->scanInProgress) {
+		qCDebug(logDesktopEntry) << "Scan already in progress, skipping";
+		return;
+	}
+
+	this->scanInProgress = true;
+	auto* scanner = new DesktopEntryScanner(this);
+	QThreadPool::globalInstance()->start(scanner);
+}
+
+QString DesktopEntryManager::extractIdFromPath(const QString& path) {
+	const auto fi = QFileInfo(path);
+	const auto canon = fi.canonicalFilePath();
+	const auto actual = canon.isEmpty() ? fi.absoluteFilePath() : canon;
+	for (const auto& rootPath: DesktopEntryManager::desktopPaths()) {
+		auto root = QDir(rootPath);
+		const auto rel = root.relativeFilePath(actual);
+		if (!rel.startsWith("../")) {
+			auto relInfo = QFileInfo(rel);
+			auto id = QString();
+			if (relInfo.path() != "." && !relInfo.path().isEmpty()) {
+				id = relInfo.path();
+				id.replace('/', '-');
+				id += '-' + relInfo.completeBaseName();
+			} else {
+				id = relInfo.completeBaseName();
+			}
+			return id;
+		}
+	}
+	return fi.completeBaseName();
+}
+
+const QStringList& DesktopEntryManager::desktopPaths() {
+	static const auto paths = []() {
+		auto dataPaths = QStringList();
+
+		auto dataHome = qEnvironmentVariable("XDG_DATA_HOME");
+		if (dataHome.isEmpty() && qEnvironmentVariableIsSet("HOME"))
+			dataHome = qEnvironmentVariable("HOME") + "/.local/share";
+		if (!dataHome.isEmpty()) dataPaths.append(dataHome + "/applications");
+
+		auto dataDirs = qEnvironmentVariable("XDG_DATA_DIRS");
+		if (dataDirs.isEmpty()) dataDirs = "/usr/local/share:/usr/share";
+
+		for (const auto& dir: dataDirs.split(':', Qt::SkipEmptyParts))
+			dataPaths.append(dir + "/applications");
```

Comment:
My bad for being unclear. Single line ifs/fors *only* if it doesn't cause a line break.